Notification Fragment 에 Jetpack Compose 적용

sanggggg opened the pull request.

sanggggg committed:
temp for git

sanggggg commented on app/build.gradle.kts:
```diff
@@ -172,6 +172,9 @@ dependencies {
     implementation("com.appyvet:materialrangebar:1.3")
     implementation("com.airbnb.android:lottie:3.4.0")
+
+    // Jetpack Compose and Flow and Paging?
+    implementation("androidx.paging:paging-compose:1.0.0-alpha14")
```
[nit]
위에 Compose dependency list 에 같이 적어놓으면 좋을 것 같아요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,129 @@ import android.os.Bundle
-        super.onViewCreated(view, savedInstanceState)
+    @Composable
+    fun NotificationItem(info: NotificationDto?) {
+        val subHeadingFontStyle = TextStyle(fontSize = 14.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Bold, color = colorResource(R.color.black))
+        val detailFontStyle = TextStyle(fontSize = 12.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Medium, color = colorResource(R.color.black))
```
코스트가 얼마나 클 지는 모르겠지만
이렇게 되면 recomposition 일어날 때 마다 TextStyle 인스턴스가 생길 것 같아서
외부에 한번만 생성해 두고 쓰는 형태가 좋을 것 같습니다

sanggggg replied:
지금 깔끔하게 디자인이 시스템화 되어있지는 않은 것 같지만
이런 Typo 들도 SNUTTTheme 안에서 Typo 에 정의되어 있는 것을 가져다 쓰는 형태가 좋을 것 같습니다.
다크모드를 고려하고 리팩토링 하기로 해서

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,129 @@ import android.os.Bundle
-                adapter.submitData(lifecycle, it)
+        Column(modifier = Modifier.padding(all = 16.dp)) {
+            Row {
+                Image(
+                    painter = painterResource(R.drawable.ic_refresh),
+                    contentDescription = "Message",
+                    modifier = Modifier.size(20.dp)
+                )
+                Spacer(modifier = Modifier.width(10.dp))
+                Column {
+                    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
+                        Text(text = typeList[info?.type ?: 0], style = subHeadingFontStyle)
+                        Text(text = info?.createdAt?.substring(0, 10) ?: "", style = detailFontStyle, color = colorResource(R.color.created_at))
+                    }
+                    Spacer(modifier = Modifier.height(7.dp))
+                    Text(text = info?.message ?: "", style = detailFontStyle)
```
이제 실제 develop 으로 합쳐지기 위해서 기존 로직을 가져오면 좋을 것 같습니다.
(drawable, typeList, createdAt 포매팅 등)
연습 프로젝트 작성 시 간단히 넘어간 부분들 fix 부탁드립니다!

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,129 @@ import android.os.Bundle
+        val items: LazyPagingItems<NotificationDto> = notifications.collectAsLazyPagingItems()
+        CompositionLocalProvider(LocalOverScrollConfiguration provides null) {
+            LazyColumn {
+                items(items) { notification ->
+                    for (i: Int in 1..10) NotificationItem(notification)
+                }
```
이렇게 10개를 한번에 박아버리면 LazyColumn 의 효과를 받고 있지 않는 형태일 것 같아요.
paging LazyColumn 의 경우에 새로 로드하냐 마냐를 판단하는 기준을 각 items 안의 item 하나 하나를 단위로 판별할 텐데, 하나의 아이템에서 10개를 로드하니 실제 LazyColumn 의 최적화가 적용되지 않을 것 같은 느낌입니다 👀 

> 연습프로젝트기에 간단히 작성하신 것으로 보이는데, 만약 앞으로 이런 모킹 (10개를 보여줘야함) 같은게 필요하다면 앞으로는 ViewModel 단에서 Flow<PagingData> 에게 이뤄져서 보이는게 좋을 것 같습니다.
> ViewModel 에서 최대한 무엇을 보여줄지를 관리해야 View 는 모킹이던 실제던 원하는 동작대로 동작함을 보장할 수 있을 것 같아요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,129 @@ import android.os.Bundle
+                    modifier = Modifier
+                        .size(30.dp)
+                        .clickable(
+                            interactionSource = remember { MutableInteractionSource() },
```
질문)
InteractionSource 추가한 이유가 있으신가요?
(remember 된 interactionSource 가 백 버튼이 아닌 다른 곳에서 사용되지 않는 것 같아서 여쭤봅니다)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsViewModel.kt:
```diff
@@ -2,21 +2,19 @@ package com.wafflestudio.snutt2.views.logged_in.notifications
 class NotificationsViewModel @Inject constructor(
-    private val notificationsRepository: NotificationsRepository
+    private val notificationRepository: NotificationsRepository,
+    private val apiOnError: ApiOnError
```
[nit]
apiOnError 는 필요없는 디펜던시 인 것 같아요

JuTaK97 committed:
NotificationFragment JetpackCompose 피드백 반영

sanggggg committed:
Merge branch 'develop' into jutak97/compose-notification-fragment

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsViewModel.kt:
```diff
@@ -5,18 +5,14 @@ import androidx.paging.PagingData
 @HiltViewModel
 class NotificationsViewModel @Inject constructor(
-    private val notificationsRepository: NotificationsRepository
+    private val notificationRepository: NotificationsRepository
```
[nit]
```suggestion
    private val notificationsRepository: NotificationsRepository
```

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,171 @@ import android.os.Bundle
+                                val format: DateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+                                format.timeZone = TimeZone.getTimeZone("UTC")
+                                val date1 = format.parse(info?.createdAt) ?: Date()
+                                val date2 = Date()
 
-        vm.getNotifications()
-            .bindUi(this) {
-                adapter.submitData(lifecycle, it)
+                                val diff = date2.time - date1.time
+                                val hours = diff / (1000 * 60 * 60)
+                                val days = hours / 24
+                                when {
+                                    days > 0 -> {
+                                        DateFormat.getDateInstance().format(date1)
+                                    }
+                                    hours > 0 -> {
+                                        "$hours 시간 전"
+                                    }
+                                    else -> {
+                                        "방금"
+                                    }
+                                }
+                            } catch (e: ParseException) {
+                                Timber.e("notification created time parse error!")
+                                "-"
+                            },
```
시간 가공의 영역은 util 로 빼기

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,171 @@ import android.os.Bundle
+                    Spacer(modifier = Modifier.height(2.dp))
+                    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
+                        when (info?.type) {
+                            0 -> context?.getString(R.string.notifications_noti_warning)
+                            1 -> context?.getString(R.string.notifications_noti_add)
+                            2 -> context?.getString(R.string.notifications_noti_update)
+                            3 -> context?.getString(R.string.notifications_noti_delete)
```
stringResources(R.string.notifcations_noti_warning) 처럼 사용하기

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,171 @@ import android.os.Bundle
-    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
-        super.onViewCreated(view, savedInstanceState)
+    @Composable
+    fun NotificationItem(info: NotificationDto?) {
```
activity 의 method 가 아닌 순수함수로 다른 파일에 분리해서 컴포넌트를 정의하고,
그 파일에서 잘 모킹된 값으로 Preview 를 보여주는 건 어떨까요?

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,171 @@ import android.os.Bundle
+                    3 -> painterResource(R.drawable.ic_trash)
+                    else -> null
+                }?.let {
+                    Image(painter = it, contentDescription = "Message", modifier = Modifier.size(20.dp))
```
stringResource 로 빼는게 접근성 대응에 좋다
다만... 좀 더 고민해보자

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt:
```diff
@@ -4,64 +4,171 @@ import android.os.Bundle
+                    contentDescription = "back button",
+                    modifier = Modifier
+                        .size(30.dp)
+                        .clickable(
+                            interactionSource = MutableInteractionSource(),
+                            onClick = { findNavController().popBackStack() },
+                            indication = null
+                        )
```
꼭 필요없는 overload 함수를 쓴 것 같아서 수정 부탁드립니다

sanggggg reviewed (commented):
에러 / 빈 화면일때 Placeholder 도 반영 부탁드립니다

JuTaK97 committed:
2차 피드백 반영

JuTaK97 committed:
TopBar 함수 범용성

JuTaK97 committed:
lint 해결

sanggggg commented on app/src/main/res/layout/fragment_notifications.xml:
```diff
@@ -45,14 +45,15 @@
             android:layout_height="match_parent"
             android:gravity="center"
-            tools:visibility="gone">
+            tools:visibility="visible">
```
해당 파일은 삭제 해도 괜찮을 것 같아요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/lib/data/SNUTTStringUtils.kt:
```diff
@@ -50,4 +55,31 @@ object SNUTTStringUtils {
         return text
     }
+
+    fun getNotificationTime(info: NotificationDto?): String {
```
info 를 안 받았을 때는 아예 `-` 같은 형태로 값을 보여주는게 좋을 것 같아요.
getNotificationTime 은 파라미터로 논 널 값을 받고, 이 함수를 부르는 곳에서 info 가 null 이라면 `-` 를 UI 에 보여주는 식으로 대응하면 좋을 것 같습니다.

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsComposable.kt:
```diff
@@ -0,0 +1,190 @@
+    } else {
+        LazyColumn {
+            items(items) { notification ->
+                NotificationItem(notification)
```
fyi

요거 깔끔하게 대응하려면 [NotificationItem 에서 null 도 place 홀더 대응을 하라길래](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:paging/paging-compose/src/main/java/androidx/paging/compose/LazyPagingItems.kt;l=239-242;drc=2dfb23550477e600bf4db33bf4265ac12bfce579) 왜 notification 의 타입이 nullable 한지 궁금해서 찾아봤는데
[LazyPagingItems 에서 pagingDiffer 의 연산이 async 하기 때문](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:paging/paging-common/src/main/kotlin/androidx/paging/PagingDataDiffer.kt;l=177;drc=21534897538717811f9b0da9926e13e63b664911)에 최초에 null 이었다가, differ 연산이 끝나면 다시 null 이 아닌 값으로 업뎅이트 하기 때문에, null 인 경우 UI 를 순간 대응할 필요가 있어서 존재하는 것 같네요, 요런 부분 생각하면서 NotificationItem 이 null 일때 어떻게 하면 예쁘게 보일까 고민해도 좋을 것 같습니다

JuTaK97 replied:
https://github.com/wafflestudio/SNUTT-android/commit/23b18bc7f998be9c739bfcb0ebfcbb73665d7d4a

sanggggg committed:
Merge remote-tracking branch 'origin/develop' into jutak97/compose-notification-fragment

JuTaK97 committed:
getNotificationItem NULL 값 대응

https://github.com/wafflestudio/SNUTT-android/pull/48#discussion_r846647135

JuTaK97 committed:
Repository 레이어 리팩토링 (#49) 반영

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsComposable.kt:
```diff
@@ -0,0 +1,190 @@
+    Font(R.font.spoqa_han_sans_bold, FontWeight.Bold),
+)
+
+private val subHeadingFontStyle = TextStyle(fontSize = 14.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Bold)
+private val subHeading2FontStyle = TextStyle(fontSize = 14.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Medium)
+private val detailFontStyle = TextStyle(fontSize = 12.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Medium)
+private val titleStyle = TextStyle(fontSize = 17.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Bold)
+private val placeholderTitle = TextStyle(fontSize = 18.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Bold)
+private val placeholderDetail = TextStyle(fontSize = 14.sp, fontFamily = spoqaHanSans, fontWeight = FontWeight.Medium)
```
Typo 로 빼기!

JuTaK97 committed:
TextStyle들 Typography로 분리

JuTaK97 committed:
Ripple 효과 제거

NotificationFragment에서 뒤로 가기 버튼의 ripple effect를 제거한다.

indication=null로 하기 위해 clickable의 다른 overload를 사용해야 하고, interactionSource param도 있어야 한다.
